src/MCPClient/lib/clientScripts/verify_checksum.py

```diff
@@ -0,0 +1,276 @@
+        try:
+            return self._call("--version")[0]
+        except subprocess.CalledProcessError:
+            return self.COMMAND
+
+    def get_command_detail(self):
+        """Provide some way for the user to get information out of the class to
+        write METS/PREMIS provenance information.
+        """
+        if not self.command_called:
+            raise PREMISFailure()
+        return 'program="{}"; version="{}"'.format(
+            " ".join(self.command_called), self.version()
+        )
+
+    @staticmethod
+    def _decode(out):
+        """Decode string output in Py2 or Py3 and return a list of lines to be
+        parsed elsewhere.
+        """
+        try:
+            return str(out, "utf8").split("\n")
+        except TypeError:
+            return out.decode("utf8").split("\n")
+
+    @staticmethod
+    def get_ext(path):
+        """Return the extension of the checksum file provided"""
+        ext = os.path.splitext(path)[1]
+        if not ext:
+            return path
+        return ext.replace(".", "")
+
+    @staticmethod
+    def _count_lines(path):
+        """Count the number of lines in a checksum file."""
+        count = 0
+        with open(path) as hashfile:
+            for count, _ in enumerate(hashfile):
+                pass
+        # Negate zero-based count.
+        return count + 1
+
+    @staticmethod
+    def _count_files(path):
+        """Walk the directories on a given path and count the number of files.
+        """
+        return sum([len(files) for _, _, files in os.walk(path)])
+
+
+def write_premis_event_per_file(transfer_uuid, event_detail):
+    """Write a PREMIS event for every file in the transfer on success."""
+    group_type = "transfer_id"
+    event_type = "fixity_check"
+    event_outcome = "pass"
+    # Retrieve file UUIDs from the database.
+    kwargs = {"removedtime__isnull": True, group_type: transfer_uuid}
+    with transaction.atomic():
+        file_uuids = File.objects.filter(**kwargs).values_list("uuid")
+        if not file_uuids:
+            raise PREMISFailure()
+        for (fileUUID,) in file_uuids:
+            insertIntoEvents(
+                fileUUID=fileUUID,
+                eventType=event_type,
+                eventDateTime=datetime.datetime.now()
+                .replace(microsecond=0)
+                .isoformat(),
+                eventDetail=event_detail,
+                eventOutcome=event_outcome,
+                eventOutcomeDetailNote="",
+            )
+
+
+def run_hashsum_commands(job):
+    """Run hashsum commands and generate a cumulative return code."""
+    transfer_dir = None
+    transfer_uuid = None
+    try:
+        transfer_dir = job.args[1]
+        transfer_uuid = job.args[2]
+    except IndexError:
+        logger.error("Cannot access expected module arguments: %s", job.args)
+        return 1
+    ret = 0
+    for hashfile in Hashsum.COMMANDS:
+        os.chdir(transfer_dir)
```

Comment:
Why can't the logic of the script work off the `transfer_dir` joining absolute paths? I also noticed @sevein [raised this as an issue](https://github.com/artefactual/archivematica/pull/1333#discussion_r253208026).

Reply:
I might be missing something, and I probably wasn't clear enough in my response to Sevein, but as far as my reading and tests have gone, as far as I can work out, the check function of the tool we are looking to replace the `md5deep`-like commands with functions relative to the digest listing. 

This doesn't give us many options:

* Users create an absolute path up front.
* We run the tool in the script as we are.
* We manipulate paths in the digest and find a way of comparing an absolute path with a checksum? (this looks like our own custom checksum comparison script at this point, or finding another tool)
* Find a way of running the tool from a staging location? 

The paths we have in the microservice are:

* The path the script is executed from (`cwd`): `/`
* The path we need to be on to run the `--check` comparison: `/var/archivematica/sharedDirectory/currentlyProcessing/cs1-456358ec-5c28-45d7-842c-7e9c2cca6cfa/`

And trying this with our sample data when the objects are expected to be at: `/home/ross-spencer/git/artefactual-labs/am/src/archivematica-sampledata/TestTransfers/fixityCheckShouldFail/objects`:

```
ross-spencer@artefactual:~$ md5sum -c /home/ross-spencer/git/artefactual-labs/am/src/archivematica-sampledata/TestTransfers/fixityCheckShouldFail/metadata/checksum.md5
md5sum: ./WFPC01.GIF: No such file or directory
./WFPC01.GIF: FAILED open or read
md5sum: ./oakland03.jp2: No such file or directory
./oakland03.jp2: FAILED open or read
md5sum: ./Nemastylis_geminiflora_Flower.PNG: No such file or directory
./Nemastylis_geminiflora_Flower.PNG: FAILED open or read
md5sum: WARNING: 3 listed files could not be read
```

As I noted to Sevein, if this is a blocker we need to find an alternative. And/or:

* Figure out how to use the tool the way that works better for us. 
* Understand and prove the risk that might be caused by changing the directory and setting it back (some additional information on this would help me). 
* Understand what other alternatives we have for this function in Archivematica.

Reply:
> Understand and prove the risk that might be caused by changing the directory and setting it back (some additional information on this would help me).

I think that there's very little risk today because client scripts are executed in single-threaded context since AM 1.8.
However I think it's better to write thread-safe code where possible because that doesn't lock us in a particular execution model. E.g. if we started using threads again, context switches will happen and the use of `chdir` will likely cause race conditions.

I'm not able to prove it formally but here's a very quick example that illustrates the issue:

```python
>>> import os
>>> import threading
>>> 
>>> def change_chdir():
...     os.chdir("/etc")
... 
>>> os.chdir("/")
>>> threading.Thread(target=change_chdir).start()
>>> print(os.getcwd())
/etc
```

This example is non-deterministic as the threads are going to be scheduled by the operative system. I think that makes clearer that the context in which your code is executed is unpredictable.

> Understand what other alternatives we have for this function in Archivematica.

If all you care about is the state of the child process, you could try passing `cwd` to `subprocess.check_output`.

https://docs.python.org/2/library/subprocess.html#subprocess.Popen
> If `cwd` is not `cwd`, the child’s current directory will be changed to `cwd` before it is executed. Note that this directory is not considered when searching the executable, so you can’t specify the program’s path relative to `cwd`.

Reply:
Thanks @sevein that was really helpful. I've made the `check_output` change here; https://github.com/artefactual/archivematica/pull/1333/commits/cde917bc657f2b4817e991322345a85b8bbe0882 hopefully that's better for this microservice.